src/nifs/vanilla.rs

```diff
@@ -93,28 +91,20 @@ impl<C: CurveAffine> VanillaFS<C> {
             W2s: &W2.W,
         };
 
-        let normalized = S.poly.fold_transform(offset, S.num_fold_vars());
-        let r_index = normalized.num_challenges() - 1;
-        let degree = S.poly.degree_for_folding(offset);
-        let cross_terms: Vec<Vec<C::ScalarExt>> = (1..degree)
-            .map(|k| {
-                normalized.coeff_of(
-                    ColumnIndex::Challenge {
-                        column_index: r_index,
-                    },
-                    k,
-                )
-            })
-            .map(|multipoly| {
-                (0..num_row)
-                    .into_par_iter()
-                    .map(|row| data.eval(&multipoly, row))
-                    .collect::<Result<Vec<C::ScalarExt>, EvalError>>()
-            })
-            .collect::<Result<Vec<Vec<C::ScalarExt>>, EvalError>>()?;
-        let cross_term_commits: Vec<C> =
-            cross_terms.iter().map(|v| ck.commit(v).unwrap()).collect();
-        Ok((cross_terms, cross_term_commits))
+        itertools::process_results(
+            data.eval_grouped(&GroupedPoly::new(
```

Comment:
we should have a general evaluation method for any Expression<F>. Here the expressions are cross terms. However, in IVC::verify, we have to check `is_sat` and `is_sat_relaxed` which is evaluation on the original custom gates expressions.

Reply:
I'm conversely thinking of storing not an `Expression` but an already grouped `GroupedPoly` (as is done now with `Multipoly`), so as not to repeat the execution of the symbolic part several times. 

Then `eval_grouped(&grouped_poly)` would be just fine.

Reply:
`is_sat` draft of API implemented